File: src/mcp/embedAndUpload.ts
import { CohereClient } from 'cohere-ai';
import { QdrantClient } from '@qdrant/js-client-rest';
import { v4 as uuidv4 } from 'uuid';
import dotenv from 'dotenv';

dotenv.config();


const cohere = new CohereClient({
  token: process.env.COHERE_API_KEY!,
});

const qdrant = new QdrantClient({
  url: process.env.QDRANT_URL!,
  apiKey: process.env.QDRANT_KEY!,
});

const COLLECTION_NAME = 'document-qna';
const MAX_BATCH_SIZE = 96;

function chunkArray<T>(array: T[], chunkSize: number): T[][] {
  const result: T[][] = [];
  for (let i = 0; i < array.length; i += chunkSize) {
    result.push(array.slice(i, i + chunkSize));
  }
  return result;
}

export async function embedAndUploadChunks(chunks: string[]) {
  console.log("📦 Total chunks:", chunks.length);
  const chunkedChunks = chunkArray(chunks, MAX_BATCH_SIZE);

  for (let i = 0; i < chunkedChunks.length; i++) {
    const batch = chunkedChunks[i];
    console.log(`🔹 Processing batch ${i + 1}/${chunkedChunks.length} (size: ${batch.length})`);

    const embed = await cohere.v2.embed({
      texts: batch,
      model: 'embed-v4.0',
      inputType: 'search_document',
    });

    const embeddings = embed.embeddings['float'];

    if (!embeddings || embeddings.length !== batch.length) {
      throw new Error("❌ Embedding failed or mismatched for batch.");
    }

    const points = batch.map((chunk, index) => ({
      id: uuidv4(),
      vector: embeddings[index],
      payload: { text: chunk },
    }));

    const uploadResponse = await qdrant.upsert(COLLECTION_NAME, {
      wait: true,
      batch: {
        vectors: points.map(p => p.vector),
        ids: points.map(p => p.id),
        payloads: points.map(p => p.payload),
      },
    });

    console.log(`✅ Batch ${i + 1} uploaded successfully.`);
  }

  console.log("🎉 All chunks embedded and uploaded to Qdrant.");
}

File: src/mcp/embed_and_upload.py
import os
import uuid
import cohere
from qdrant_client import QdrantClient
from qdrant_client.models import PointStruct


def chunk_array(array: list, chunk_size: int) -> list[list]:
    """Split an array into chunks."""
    result = []
    for i in range(0, len(array), chunk_size):
        result.append(array[i:i + chunk_size])
    return result


def embed_and_upload_chunks(chunks: list[str]) -> None:
    """Embed chunks using Cohere and upload to Qdrant."""
    cohere_key = os.getenv("COHERE_API_KEY")
    qdrant_url = os.getenv("QDRANT_URL")
    qdrant_key = os.getenv("QDRANT_KEY")

    co = cohere.Client(api_key=cohere_key)
    qdrant = QdrantClient(url=qdrant_url, api_key=qdrant_key)

    collection_name = "document-qna"
    max_batch_size = 96

    print(f"📦 Total chunks: {len(chunks)}")
    chunked_chunks = chunk_array(chunks, max_batch_size)

    for i, batch in enumerate(chunked_chunks):
        print(f"🔹 Processing batch {i + 1}/{len(chunked_chunks)} (size: {len(batch)})")

        embed_response = co.embed(
            texts=batch,
            model="embed-v4.0",
            input_type="search_document"
        )

        embeddings = embed_response.embeddings

        if not embeddings or len(embeddings) != len(batch):
            raise Exception("❌ Embedding failed or mismatched for batch.")

        points = [
            PointStruct(
                id=str(uuid.uuid4()),
                vector=embeddings[idx],
                payload={"text": chunk}
            )
            for idx, chunk in enumerate(batch)
        ]

        qdrant.upsert(
            collection_name=collection_name,
            points=points,
            wait=True
        )

        print(f"✅ Batch {i + 1} uploaded successfully.")

    print("🎉 All chunks embedded and uploaded to Qdrant.")
